rust/storage/src/admissioncontrolleds3.rs

```diff
@@ -111,6 +173,42 @@ impl AdmissionControlledS3Storage {
                 return Err(AdmissionControlledS3StorageError::S3GetError(e));
             }
         }
+        // Permit gets dropped here due to RAII.
+    }
+
+    pub async fn get_parallel(
+        &self,
+        key: String,
+    ) -> Result<Arc<Vec<u8>>, AdmissionControlledS3StorageError> {
+        // If there is a duplicate request and the original request finishes
+        // before we look it up in the map below then we will end up with another
+        // request to S3.
+        let future_to_await;
+        {
+            let mut requests = self.outstanding_requests.lock();
+            let maybe_inflight = requests.get(&key).map(|fut| fut.clone());
```

Comment:
just so I understand, if there's an inflight request that was created with `.get()` and then another part of the system calls `.get_parallel()` for the same key, the `.get_parallel()`will dedupe to the original sequential `.get()`?

Reply:
It will dedup. Although I am not sure how that can happen in practice for the same key